codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -230,19 +268,50 @@ private int connectionWindowSize() {
         return connectionState().windowSize();
     }
 
+    private int minUsableChannelBytes() {
+        assert ctx != null;
+        // The current allocation algorithm values "fairness" and doesn't give any consideration to "goodput". It
+        // is possible that 1 byte will be allocated to many streams. In an effort to try to make "goodput"
+        // reasonable with the current allocation algorithm we have this "cheap" check up front to ensure there is
+        // an "adequate" amount of connection window before allocation is attempted. This is not foolproof as if the
+        // number of streams is >= this minimal number then we will have the issue, but the idea is to narrow the
+        // circumstances in which this can happen without rewriting the allocation algorithm.
+        return Math.max(ctx.channel().config().getWriteBufferHighWaterMark() >>> 1, 1 << 15);
+    }
+
+    private int maxUsableChannelBytes() {
+        if (ctx == null) {
+            return 0;
+        }
+
+        int channelWritableBytes = (int) Math.min(Integer.MAX_VALUE, ctx.channel().bytesBeforeUnWritable());
+        return channelWritableBytes >= minUsableChannelBytes() ?
+                Math.min(connectionState().windowSize(), channelWritableBytes) : 0;
```

Comment:
Going a little over is ok, we just want to be in the ballpark ... and even if we're over by 32k that should be fine. This is a tradeoff so that we always try to write a reasonably sized chunk so long as the channel is writable.